mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt

```diff
@@ -0,0 +1,1020 @@
+        // Handle hamburger menu
+        if (drawerToggle.onOptionsItemSelected(item)) {
+            return true
+        }
+
+        // Handle menu items
+        return when (item.itemId) {
+            R.id.mainmenu_voice_recognition -> {
+                launchVoiceRecognition()
+                true
+            }
+            else -> super.onOptionsItemSelected(item)
+        }
+    }
+
+    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        Log.d(TAG, "onActivityResult() requestCode = $requestCode, resultCode = $resultCode")
+        when (requestCode) {
+            SETTINGS_REQUEST_CODE -> {
+                if (data == null) {
+                    return
+                }
+                if (data.getBooleanExtra(PreferencesActivity.RESULT_EXTRA_SITEMAP_CLEARED, false)
+                        && connection != null && serverProperties != null) {
+                    val sitemap = selectConfiguredSitemapFromList()
+                    if (sitemap != null) {
+                        openSitemap(sitemap)
+                    } else {
+                        showSitemapSelectionDialog()
+                    }
+                }
+                if (data.getBooleanExtra(PreferencesActivity.RESULT_EXTRA_THEME_CHANGED, false)) {
+                    recreate()
+                }
+            }
+            WRITE_NFC_TAG_REQUEST_CODE -> Log.d(TAG, "Got back from Write NFC tag")
+        }
+    }
+
+    public override fun onSaveInstanceState(savedInstanceState: Bundle) {
+        Log.d(TAG, "onSaveInstanceState()")
+        isStarted = false
+        with (savedInstanceState) {
+            putParcelable("serverProperties", serverProperties)
+            putParcelable("sitemap", selectedSitemap)
+            putBoolean("isSitemapSelectionDialogShown", sitemapSelectionDialog?.isShowing == true)
+            putString("controller", controller.javaClass.canonicalName)
+            putInt("connectionHash", connection?.hashCode() ?: -1)
+            controller.onSaveInstanceState(this)
+        }
+        super.onSaveInstanceState(savedInstanceState)
+    }
+
+    private fun onNotificationSelected(intent: Intent) {
+        Log.d(TAG, "onNotificationSelected()")
+        // mPendingOpenedNotificationId being non-null is used as trigger for
+        // opening the notifications page, so use a dummy if it's null
+        pendingOpenedNotificationId = intent.getStringExtra(EXTRA_PERSISTED_NOTIFICATION_ID).orEmpty()
+        openNotificationsPageIfNeeded()
+    }
+
+    fun onWidgetSelected(linkedPage: LinkedPage, source: WidgetListFragment) {
+        Log.d(TAG, "Got widget link = ${linkedPage.link}")
+        controller.openPage(linkedPage, source)
+    }
+
+    fun updateTitle() {
+        val title = controller.currentTitle
+        setTitle(title ?: getString(R.string.app_name))
+        drawerToggle.isDrawerIndicatorEnabled = !controller.canGoBack()
+    }
+
+    override fun onBackPressed() {
```

Comment:
IMO override methods aka "Android framework methods" should be at the beginning of a class and all "custom/own" functions after them.